Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,314 @@
+    let eventLoops = self.group.makeIterator()
+    return eventLoops.map { eventLoop in
+      // We're creating a retain cycle here as each pool will reference the manager and the per-pool
+      // state will hold the pool which will in turn be held by the pool manager. That's okay: when
+      // the pool is shutdown the per-pool state and in turn each connection pool will be dropped.
+      // and we'll break the cycle.
+      return ConnectionPool(
+        eventLoop: eventLoop,
+        maxWaiters: configuration.maxWaiters,
+        reservationLoadThreshold: configuration.loadThreshold,
+        assumedMaxConcurrentStreams: configuration.assumedMaxConcurrentStreams,
+        channelProvider: configuration.channelProvider,
+        streamLender: self,
+        logger: logger
+      )
+    }
+  }
+
+  // MARK: Stream Creation
+
+  /// A future for a `Channel` from a managed connection pool. The `eventLoop` indicates the loop
+  /// that the `Channel` is running on and therefore which event loop the RPC will use for its
+  /// transport.
+  internal struct PooledStreamChannel {
+    /// The future `Channel`.
+    var futureResult: EventLoopFuture<Channel>
+
+    /// The `EventLoop` that the `Channel` is using.
+    var eventLoop: EventLoop {
+      return self.futureResult.eventLoop
+    }
+  }
+
+  /// Make a stream and initialize it.
+  ///
+  /// - Parameters:
+  ///   - preferredEventLoop: The `EventLoop` that the stream should be created on, if possible. If
+  ///       a pool exists running this `EventLoop` then it will be chosen over all other pools,
+  ///       irregardless of the load on the pool. If no pool exists on the preferred `EventLoop` or
+  ///       no preference is given then the pool with the most streams available will be selected.
+  ///       The `EventLoop` of the selected pool will be the same as the `EventLoop` of
+  ///       the `EventLoopFuture<Channel>` returned from this call.
+  ///   - deadline: The point in time by which the stream must have been selected. If this deadline
+  ///       is passed then the returned `EventLoopFuture` will be failed.
+  ///   - logger: A logger.
+  ///   - initializer: A closure to initialize the `Channel` with.
+  /// - Returns: A `PoolStreamChannel` indicating the future channel and `EventLoop` as that the
+  ///     `Channel` is using. The future will be failed if the pool manager has been shutdown,
+  ///     the deadline has passed before a stream was created or if the selected connection pool
+  ///     is unable to create a stream (if there is too much demand on that pool, for example).
+  internal func makeStream(
+    preferredEventLoop: EventLoop?,
+    deadline: NIODeadline,
+    logger: GRPCLogger,
+    streamInitializer initializer: @escaping (Channel) -> EventLoopFuture<Void>
+  ) -> PooledStreamChannel {
+    let reservation = self.lock.withLock {
+      self.state.reserveStream(
+        preferringPoolWithEventLoopID: preferredEventLoop.map { EventLoopID($0) }
```

Comment:
Can we lift this `preferredEventLoop.map` call out of the lock? It's nice to do as little as possible while holding that lock.